auto_dev/protocols/formatter.py

```diff
@@ -0,0 +1,247 @@
+        case ast.MapField:
+            key_type = PRIMITIVE_TYPE_MAP.get(element.key_type, element.key_type)
+            value_type = qualified_type(message, element.value_type)
+            return f"{element.name}: dict[{key_type}, {value_type}]"
+        case ast.Group | ast.Option | ast.ExtensionRange | ast.Reserved | ast.Extension:
+            msg = f"{element}"
+            raise NotImplementedError(msg)
+        case _:
+            msg = f"Unexpected message type: {element}"
+            raise TypeError(msg)
+
+
+def render(file: FileAdapter):
+    """Main function to render a .proto file."""
+
+    enums = "\n".join(render_attribute(e, file) for e in file.enums)
+    messages = "\n".join(render_attribute(e, file) for e in file.messages)
+
+    return f"{enums}\n{messages}"
+
+
+def encode_field(element, message):
+    """Render pydantic model field encoding."""
+
+    instance_attr = f"{message.name.lower()}.{element.name}"
+    if (
+        element.type in PRIMITIVE_TYPE_MAP
+        or element.type in message.enum_names
+        or element.type in message.file.enum_names
+    ):
+        value = instance_attr
+    else:  # Message
+        qualified = qualified_type(message, element.type)
+        if element.cardinality == FieldCardinality.REPEATED:
+            return f"for item in {instance_attr}:\n" f"    {qualified}.encode(proto_obj.{element.name}.add(), item)"
+        if element.cardinality == FieldCardinality.OPTIONAL:
+            return (
+                f"if {instance_attr} is not None:\n"
+                f"    temp = proto_obj.{element.name}.__class__()\n"
+                f"    {qualified}.encode(temp, {instance_attr})\n"
+                f"    proto_obj.{element.name}.CopyFrom(temp)"
+            )
+        return f"{qualified}.encode(proto_obj.{element.name}, {instance_attr})"
+
+    match element.cardinality:
+        case FieldCardinality.REPEATED:
+            iter_items = f"for item in {value}:\n"
+            return f"{iter_items}    proto_obj.{element.name}.append(item)"
+        case FieldCardinality.OPTIONAL:
+            return f"if {instance_attr} is not None:\n    proto_obj.{element.name} = {instance_attr}"
+        case _:
+            return f"proto_obj.{element.name} = {value}"
+
+
+def render_encoder(message: MessageAdapter) -> str:
+    """Render pydantic model .encode() method."""
+
+    def encode_element(element) -> str:
+        match type(element):
+            case ast.Comment:
+                return f"# {element.text}"
+            case ast.Field:
+                return encode_field(element, message)
+            case ast.OneOf:
+                return "\n".join(
+                    f"if isinstance({message.name.lower()}.{element.name}, {PRIMITIVE_TYPE_MAP.get(e.type, e.type)}):\n    proto_obj.{e.name} = {message.name.lower()}.{element.name}"
+                    for e in element.elements
+                )
```

Comment:
This certainly is not complete yet, as I'm not using qualified_type here. We'll find out once we have such a case where a `oneof` references for example a nested message